src/java/org/apache/cassandra/index/sai/disk/v2/sortedterms/SortedTermsReader.java

```diff
@@ -0,0 +1,305 @@
+     * <p>
+     * The cursor is to be used in a single thread.
+     * The cursor is valid as long this object hasn't been closed.
+     * You must close the cursor when you no longer need it.
+     */
+    public @Nonnull Cursor openCursor() throws IOException
+    {
+        return new Cursor(termsData);
+    }
+
+    /**
+     * Allows reading the terms from the terms file.
+     * Can quickly seek to a random term by <code>pointId</code>.
+     * <p>
+     * This object is stateful and not thread safe.
+     * It maintains a position to the current term as well as a buffer that can hold one term.
+     */
+    @NotThreadSafe
+    public class Cursor implements AutoCloseable
+    {
+        private final IndexInput termsData;
+
+        // The term the cursor currently points to. Initially empty.
+        private final BytesRef currentTerm;
+
+        // The point id the cursor currently points to. -1 means before the first item.
+        private long pointId = -1;
+
+        Cursor(IndexInput termsData) throws IOException
+        {
+            this.termsData = termsData.clone();  // clone needed to get our own file pointer independently of other Cursors
+            this.termsData.seek(termsDataStartFP);
+            this.currentTerm = new BytesRef(meta.maxTermLength);
+        }
+
+        /**
+         * Returns the number of terms
+         */
+        public long count()
+        {
+            return SortedTermsReader.this.count();
+        }
+
+        /**
+         * Returns the current position of the cursor.
+         * Initially, before the first call to {@link Cursor#advance}, the cursor is positioned at -1.
+         * After reading all the items, the cursor is positioned at index one
+         * greater than the position of the last item.
+         */
+        public long pointId()
+        {
+            return pointId;
+        }
+
+        /**
+         * Returns the current term data as <code>ByteComparable</code> referencing the internal term buffer.
+         * The term data stored behind that reference is valid only until the next call to
+         * {@link Cursor#advance} or {@link Cursor#seekToPointId(long)}.
+         */
+        public @Nonnull ByteComparable term()
+        {
+            return ByteComparable.fixedLength(currentTerm.bytes, currentTerm.offset, currentTerm.length);
+        }
+
+        /**
+         * Moves the cursor to the next term and reads it into the current term buffer.
+         * <p>
+         * If there are no more available terms, clears the term buffer and the cursor's position will point to the
+         * one behind the last item.
+         * <p>
+         * This method has constant time complexity.
+         *
+         * @return true if the cursor was moved successfully, false if end of file was reached
+         * @throws IOException if a read from the terms file fails
+         */
+        public boolean advance() throws IOException
```

Comment:
`next()`

Reply:
Semantics of next() in iterators are different; if we called it "next" it would be confusing because it *does not return* the next item and does not behave the same way as iterator's next. Things that behave differently should have different names.

Reply:
> Moves the cursor to the **next** term and reads it into the current term buffer.

Reply:
By picking a random word from the method description, you're not refuting any of the two arguments I made in my previous comment. 
1. The method does not *return* the next item, so it should not be called `next`. 
2. Using a noun-based name for a method which is called for a side-effect is plain confusing.  

If you really, really insist on using word `next` let's call it `seekToNext`.

Reply:
`PostingList#advance(targetrowid)` advances to rowid.  `nextPosting` goes to the next posting.  They are usable interchangeable in the same object session.  

Disjoint method naming.

Reply:
Cool, so if you want to follow the naming convention from another class now, why your original class didn't follow it? 
For example you didn't name it `advance(targetPointId)` in the first place, but it was named `seekToPointId`. 

Let's call them `advance(pointId)` and `advanceToNext()` then?

Reply:
@mike-tr-adamson ^

Reply:
I find all the naming here confusing and some of the methods redundant but that seems to be the case for a lot of the low-level classes in SAI. If we really want to keep digging at this then:

Firstly, we have a `reset` method that is only used in tests and the `advance` method is public but is only used in tests.

The `count` & `pointId` methods really don't need to exist because aren't used outside of testing. The `term` method is only used because the `seekToPointId` method doesn't return the `currentTerm`.

From a usage perspective my preference would be for the `Cursor` to expose a single method that returns a `term` for the supplied `pointId`. If I had a preference I would name this `advanceToPointId`. I only prefer this because all the other `Cursor` implementations in Cassandra use `advance` for moving to the next item.

All the tests should test just this method and not the internals of the implementations. 

That is my 2c but I really am not that hung up on it and don't see the point in endless discussions about naming when very little of the naming in our codebase makes any sense.

Reply:
I can remove reset if we don't plan to use it. It was just trivial to implement, and we have reset in other places.

As for the split between `advance()` and `term()` - that's on purpose.
First, I just followed the design that is present in Cassandra, e.g. see `PartitionIndexIterator`. SAI is part of Cassandra and not Lucene, so I'd rather follow Cassandra conventions than Lucene conventions.

Second, there is a general widely accepted rule of clean code that methods should either mutate state or query state, but not both at the same time. 

> All the tests should test just this method and not the internals of the implementations.

These are unit tests, and the more granular tests, the easier it is to find a bug. 
Also, "internals of the implementations" is relative. A public method can be still viewed as in internal implementation of a package, a smaller package can be viewed as a internal implementation of a bigger encapsulating package, etc.